VDA-XP-Reward/contracts/VDAXPReward.sol

```diff
@@ -0,0 +1,277 @@
+import "@openzeppelin/contracts-upgradeable/utils/structs/EnumerableSetUpgradeable.sol";
+
+import { VDAVerificationContract } from "@verida/vda-verification-contract/contracts/VDAVerificationContract.sol";
+import { IVeridaDIDRegistry } from "@verida/did-registry-contract/contracts/IVeridaDIDRegistry.sol";
+import { IVDAXPReward } from "./IVDAXPReward.sol";
+import { DateTime } from "./DateTime.sol";
+
+// import "hardhat/console.sol";
+
+contract VDAXPReward is IVDAXPReward, VDAVerificationContract{
+    using EnumerableSetUpgradeable for EnumerableSetUpgradeable.AddressSet;
+
+    /** RewardToken : ERC20 contract */
+    IERC20Upgradeable internal rewardToken;
+
+    /** DIDRegistryContract : Check the claiming request is registered */
+    IVeridaDIDRegistry didRegistryContract;
+
+    /** Denominator for rate values */
+    uint32 internal rateDenominator;
+
+    /** XP - VDA conversion rate */
+    uint32 internal conversionRate; // XP to VDA token rate
+
+    /**
+     * @notice Used to check that a `signature` is used once per month
+     */
+    mapping(bytes => bool) internal isClaimedSignature;
+
+    /** Used to check that user claim only once for same `typeId` */
+    mapping(bytes => bool) internal isClaimedTypeId;
+
+    /** Used to check that the combination of `typeId` and `uniqueId` is claimed before*/
+    mapping(bytes => bool) internal isClaimedUniqueId;
+
+    /**
+     * @notice Gap for later use
+     */
+    uint256[20] private __gap;
+
+    // Custom errors
+    error InvalidValue();
+    error InvalidConversionRate();
+    error UnregisteredDIDAddress();
+    error EmptyClaimData();
+    error InvalidXP(address didAddress, string typeId, uint xp);
+    error InvalidProofTime(address didAddress, string typeId, uint16 year, uint8 month);
+    error InvalidProof(address didAddress, string typeId, uint xp, bytes signature);
+    error InsufficientTokenAmount(uint requestedAmount, uint currentAmount);
+    error DuplicatedRequest(address didAddress, string typeId, uint xp, bytes signature);
+    error DuplicatedUniqueId(address didAddress, string typeId, string uniqueId, uint xp);
+
+    function __VDAXPReward_init(IERC20Upgradeable token, IVeridaDIDRegistry didRegistry) public initializer {
+        __VDAVerificationContract_init();
+        __VDAXPReward_init_unchained(token, didRegistry);
+    }
+
+    function __VDAXPReward_init_unchained(IERC20Upgradeable token, IVeridaDIDRegistry didRegistry) internal {
+        rewardToken = token;
+        didRegistryContract = didRegistry;
+        rateDenominator = 10000000; // Set up rate from 0.000001
+    }
+
+    /**
+     * @dev See {IVDAXPReward}
+     */
+    function getTokenAddress() external virtual view override returns(address) {
+        return address(rewardToken);
+    }
+
+    /**
+     * @dev See {IVDAXPReward}
+     */
+    function getDIDRegistryAddress() external virtual view override returns(address) {
```

Comment:
Why is this needed? I would like to avoid this if at all possible.

Reply:
The `XPReward` contract is initialized with `Token` contract and `DIDRegistry` contract.
In some cases, we need to check the addresses of the `Token` and `DIDRegistry` contracts.

Reply:
As discussed, this will be removed